core/src/main/java/io/parsingdata/metal/expression/value/Fold.java

```diff
@@ -31,46 +32,46 @@
 import io.parsingdata.metal.encoding.Encoding;
 
 /**
- * Base class for {@link ValueExpression} implementations of the Fold
+ * Base class for {@link SingleValueExpression} implementations of the Fold
  * operation.
  * <p>
  * Fold has three operands: <code>values</code> (a {@link ValueExpression}),
  * <code>reducer</code> (a {@link BinaryOperator}) and <code>initial</code> (a
- * {@link ValueExpression}). First <code>initial</code> is evaluated. If it
- * does not return a single value, the final result is an empty list. Next,
+ * {@link SingleValueExpression}). First <code>initial</code> is evaluated. If it
+ * does not return a valid value, the final result is an empty list. Next,
  * <code>values</code> is evaluated and its result is passed to the abstract
  * {@link #prepareValues(ImmutableList)} method. The returned list is prefixed
  * by the value returned by evaluating <code>initial</code>. On this list, the
  * <code>reducer</code> is applied to the first two values until a single
  * value remains, which is then returned.
  */
-public abstract class Fold implements ValueExpression {
+public abstract class Fold implements SingleValueExpression {
 
     public final ValueExpression values;
     public final BinaryOperator<ValueExpression> reducer;
-    public final ValueExpression initial;
+    public final SingleValueExpression initial;
 
-    public Fold(final ValueExpression values, final BinaryOperator<ValueExpression> reducer, final ValueExpression initial) {
+    public Fold(final ValueExpression values, final BinaryOperator<ValueExpression> reducer, final SingleValueExpression initial) {
```

Comment:
Could we sometime update the shorthand to pass an expression evaluation to empty instead of these nulls? Cleans up the implementation below as well.

Reply:
This is part of #187, but I agree it's a reasonably small part and can quickly make things better.